finesse/hardware/temperature/dp9800.py

```diff
@@ -0,0 +1,238 @@
+"""This module provides an interface to DP9800 temperature readers."""
+import logging
+from datetime import datetime
+from decimal import Decimal
+
+from pubsub import pub
+from serial import Serial, SerialException
+
+from .temperature_monitor_base import TemperatureMonitorBase
+
+
+def check_data(data: bytes) -> None:
+    """Perform message integrity checks.
+
+    Check characters that we know should be constant. The message should
+    start with STX, end with NUL and contain an ETX. The exact position
+    of ETX is uncertain since it depends on the width of the BCC.
+
+    Args:
+        data: the message to check
+
+    Raises:
+        DP9800Error: Malformed message received from device
+    """
+    header_width = 2
+    value_width = 8
+    num_values = 9
+    if len(data) < (header_width + value_width * num_values):
+        raise DP9800Error("Insufficient data read")
+
+    if data[0] != 2:  # STX
+        raise DP9800Error("Start transmission character not detected")
+    if data.find(3) == -1:  # ETX
+        raise DP9800Error("End transmission character not detected")
+    if data[-1] != 0:  # NUL
+        raise DP9800Error("Null terminator not detected")
+
+
+def calculate_bcc(data: bytes) -> int:
+    """Calculate block check character.
+
+    Args:
+        data: the message to check
+
+    Returns:
+        bcc: block check character
+    """
+    bcc_chars = data[1:-2]
+    bcc = 0
+    for byte in bcc_chars:
+        bcc ^= byte
+
+    return bcc
+
+
+def parse_data(data: bytes) -> tuple[list[Decimal], str]:
+    """Parse temperature data read from the DP9800.
+
+    The sequence of bytes is translated into a list of ASCII strings
+    representing each of the temperatures, and finally into floats.
+
+    Returns:
+        vals: A list of Decimals containing the temperature values recorded
+              by the DP9800 device.
+        sysflag: string representation of the system flag bitmask
+    """
+    check_data(data)
+    bcc = calculate_bcc(data)
+    if bcc != data[-2]:
+        raise DP9800Error("BCC check failed")
+
+    try:
+        data_ascii = data.decode("ascii")
+    except UnicodeDecodeError as e:
+        raise DP9800Error(e)
+
+    vals_begin = 3
+    vals_end = 74  # assuming 9 temperatures
```

Comment:
Why 3 and 74? Need a bit of an explanation - maybe in the docstring - of the meaning of these numbers.

Reply:
I'd write `vals_end` as a negative index, i.e. so the position is relative to the end of the string. It'll be more obvious that way to whoever's reading the code.